### PR TITLE
feat(lep): engine upgrade enforcement design

### PR DESCRIPTION
**Pull request by c3y1huang**
https://github.com/longhorn/longhorn/issues/5842

### CODE REVIEW

File: enhancements/20230815-engine-upgrade-enforcement.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+The primary goal of this proposal is to enhance Longhorn's upgrade mechanism by introducing logic that prevents upgrading to Longhorn versions while there are incompatible engine images in use.
+### Non-goals [optional]
+
+`None`
+
+## Proposal
+
+This proposal focuses on preventing users from upgrading to unsupported or incompatible engine versions. This enhancement will build upon the existing pre-upgrade checks to include validation of engine version compatibility.
+
```
Comment (innobead) [inline enhancements/20230815-engine-upgrade-enforcement.md:24]: Need to consider GitOps support. ref: https://github.com/longhorn/longhorn/issues/6415

**Review by innobead — APPROVED** [review]
LGTM.

File: enhancements/20230815-engine-upgrade-enforcement.md (hunk 1 above)
Comment (innobead) [reply enhancements/20230815-engine-upgrade-enforcement.md:24]: @mantissahz is helping #6415 . Go discuss together if needed.

**Comment by innobead** [comment]
One thing that should be considered is are users able to change the engine image of a volume to an incompatible version after Longhorn upgrade, then cause the volume engine upgrade failure. or do we already have that mechanism to check the image compatibility already? If yes, then it's good.

File: enhancements/20230815-engine-upgrade-enforcement.md (hunk 1 above)
Comment (c3y1huang) [reply enhancements/20230815-engine-upgrade-enforcement.md:24]: Checked with @mantissahz , its fine to add [here](https://github.com/longhorn/longhorn-manager/blob/840dbf1509481c4fe3f062c85c525593e3db0c63/upgrade/util/util.go#L168).

**Comment by c3y1huang** [comment]
Regarding CRD rollback concern raised by @PhanLe1010 , I've verified the following scenario:
1. Applied Longhorn v1.4.1 using kubectl apply.
1. Created and attached a volume.
1. Upgraded to Longhorn v1.5.1 using kubectl apply.
1. Checked the instance managers.
```
NAME                                                  STATE     TYPE      NODE            AGE
instance-manager-r-a2e163a353734ecb3dfaaf9db617b5e2   running   replica   ip-10-0-2-212   2m10s
instance-manager-r-989d79ece6573f56da53b4fc168679f1   running   replica   ip-10-0-2-20    2m10s
instance-manager-e-361b71d234e283f1ae5bc8f82cf6f99e   running   engine    ip-10-0-2-151   2m9s
instance-manager-r-361b71d234e283f1ae5bc8f82cf6f99e   running   replica   ip-10-0-2-151   2m9s
instance-manager-1e0a30c8da417b8b897493ac2ba15307     running   aio       ip-10-0-2-20    45s
instance-manager-7f1d4c361d48a7ac7e847eeaa67a3f83     running   aio       ip-10-0-2-151   44s
instance-manager-810b3520ebb0e8c245eb9aafa63ac795     running   aio       ip-10-0-2-212   18s
```
1. Upgraded Longhorn to the fixed version, changes including:
  - Remove InstanceManager CRD spec.type
  - Engine version enforcement
1. Instance Manager After Failed Upgrade:
```
NAME                                                  STATE     TYPE   NODE            AGE
instance-manager-r-a2e163a353734ecb3dfaaf9db617b5e2   running          ip-10-0-2-212   5m46s
instance-manager-r-989d79ece6573f56da53b4fc168679f1   running          ip-10-0-2-20    5m46s
instance-manager-e-361b71d234e283f1ae5bc8f82cf6f99e   running          ip-10-0-2-151   5m45s
instance-manager-r-361b71d234e283f1ae5bc8f82cf6f99e   running          ip-10-0-2-151   5m45s
instance-manager-1e0a30c8da417b8b897493ac2ba15307     running          ip-10-0-2-20    4m21s
instance-manager-7f1d4c361d48a7ac7e847eeaa67a3f83     running          ip-10-0-2-151   4m20s
instance-manager-810b3520ebb0e8c245eb9aafa63ac795     running          ip-10-0-2-212   3m54s
```
1. Applied Longhorn v1.5.1 again.
1. Instance Manager after rollback to v1.5.1.
```
NAME                                                  STATE     TYPE      NODE            AGE
instance-manager-r-a2e163a353734ecb3dfaaf9db617b5e2   running   replica   ip-10-0-2-212   7m19s
instance-manager-r-989d79ece6573f56da53b4fc168679f1   running   replica   ip-10-0-2-20    7m19s
instance-manager-e-361b71d234e283f1ae5bc8f82cf6f99e   running   engine    ip-10-0-2-151   7m18s
instance-manager-r-361b71d234e283f1ae5bc8f82cf6f99e   running   replica   ip-10-0-2-151   7m18s
instance-manager-1e0a30c8da417b8b897493ac2ba15307     running   aio       ip-10-0-2-20    5m54s
instance-manager-7f1d4c361d48a7ac7e847eeaa67a3f83     running   aio       ip-10-0-2-151   5m53s
instance-manager-810b3520ebb0e8c245eb9aafa63ac795     running   aio       ip-10-0-2-212   5m27s
```

**Comment by c3y1huang** [comment]
Summary after discussing with @innobead regarding the possibility of a pre-upgrade version check.

Users can check the engine image version in UI and CRs against the compatible version in the release note.
- The versions are already provided EngineImage CRD.
```
- apiVersion: longhorn.io/v1beta2
  kind: EngineImage
  metadata:
    creationTimestamp: "2023-08-16T03:51:12Z"
    finalizers:
    - longhorn.io
    generation: 1
    labels:
      longhorn.io/component: engine-image
      longhorn.io/engine-image: ei-74783864
      longhorn.io/managed-by: longhorn-manager
    name: ei-74783864
    namespace: longhorn-system
    resourceVersion: "166918"
    uid: 4f415b74-64d7-4ab1-87fd-bb11bdd65bc5
  spec:
    image: longhornio/longhorn-engine:v1.5.1
  status:
    buildDate: "2023-07-19T05:05:37+00:00"
    cliAPIMinVersion: 3
    cliAPIVersion: 8
    conditions:
    - lastProbeTime: ""
      lastTransitionTime: "2023-08-16T03:51:20Z"
      message: Engine image ei-74783864 (longhornio/longhorn-engine:v1.5.1) is fully
        deployed on all ready nodes
      reason: ""
      status: "True"
      type: ready
    controllerAPIMinVersion: 3
    controllerAPIVersion: 4
    dataFormatMinVersion: 1
    dataFormatVersion: 1
    gitCommit: 0d8e42d0c74b1d0430e56efa1e8247c55f34cef3
    noRefSince: "2023-08-16T03:51:13Z"
    nodeDeploymentMap:
      ip-10-0-2-20: true
      ip-10-0-2-151: true
      ip-10-0-2-212: true
    ownerID: ip-10-0-2-212
    refCount: 0
    state: deployed
    version: v1.5.1
kind: List
metadata:
  resourceVersion: ""
```
- UI (TBU).

**Pull request merged**